cpp/src/barretenberg/stdlib/primitives/logic/logic.cpp

```diff
@@ -0,0 +1,94 @@
+#include "logic.hpp"
+
+#include "../composers/composers.hpp"
+#include "../plookup/plookup.hpp"
+
+namespace plonk {
+namespace stdlib {
+
+/**
+ * @brief Constrants a logical AND or XOR over a variable number of bits.
+ *
+ * Defaults to basic Composer method if not using plookup-compatible composer
+ *
+ * @tparam Composer
+ * @param a
+ * @param b
+ * @param num_bits
+ * @param is_xor_gate
+ * @return field_t<Composer>
+ */
+template <typename Composer>
+field_t<Composer> logic<Composer>::create_logic_constraint(field_pt& a, field_pt& b, size_t num_bits, bool is_xor_gate)
+{
+    // can't extend past field size!
+    ASSERT(num_bits < 254);
+    if (a.is_constant() && b.is_constant()) {
+        uint256_t a_native(a.get_value());
+        uint256_t b_native(b.get_value());
+        uint256_t c_native = is_xor_gate ? (a_native ^ b_native) : (a_native & b_native);
+        return field_t<Composer>(c_native);
+    }
+    if (a.is_constant() && !b.is_constant()) {
+        Composer* ctx = b.get_context();
+        uint256_t a_native(a.get_value());
+        field_t<Composer> a_witness = field_pt::from_witness_index(ctx, ctx->put_constant_variable(a_native));
+        return create_logic_constraint(a_witness, b, num_bits, is_xor_gate);
+    }
+    if (!a.is_constant() && b.is_constant()) {
+        Composer* ctx = a.get_context();
+        uint256_t b_native(b.get_value());
+        field_pt b_witness = field_pt::from_witness_index(ctx, ctx->put_constant_variable(b_native));
+        return create_logic_constraint(a, b_witness, num_bits, is_xor_gate);
+    }
+    if constexpr (Composer::type == ComposerType::PLOOKUP) {
+        Composer* ctx = a.get_context();
+
+        const size_t num_chunks = (num_bits / 32) + ((num_bits % 32 == 0) ? 0 : 1);
+        uint256_t left(a.get_value());
+        uint256_t right(b.get_value());
+
+        field_pt res(ctx, 0);
```

Comment:
I think this may need additional constraints to be secure. We have a conversation going elsewhere on that point.

Reply:
@Globallager ^